--- Makefile.orig	2011-06-15 02:41:30.000000000 +1200
+++ Makefile	2011-06-29 14:23:15.467203755 +1200
@@ -3,10 +3,10 @@
 
 .include <bsd.own.mk>
 
-.PATH: ${.CURDIR}/../../usr.bin/wall
+BINDIR=	%%PREFIX%%/sbin
 
 PROG=	syslogd
-MAN=	syslog.conf.5 syslogd.8
+MAN=
 SRCS=	syslogd.c ttymsg.c
 
 DPADD=	${LIBUTIL}
@@ -18,6 +18,6 @@
 CFLAGS+= -DINET6
 .endif
 
-CFLAGS+= -I${.CURDIR}/../../usr.bin/wall
+CFLAGS+= -I%%LOCALBASE%%/include -L%%LOCALBASE%%/lib
 
 .include <bsd.prog.mk>
